content/posts/distributed-business-processes.md

```diff
@@ -0,0 +1,80 @@
+
+## Architecture is not an exact science
+
+There are two types of architects our there. There are ones that will tell you exactly what patterns and technologies you should use even before you have a chance to explain them what problem your system is solving. There are also ones that will tell you either "it depends" or "you milage may vary" even if you have given them exact specification of the application. That says a lot. Software architecture is not an exact science. There are hardly any rules justified by solid research. Most of it boils down to which conference you have attended recently. But do not despair. It turns out that if you look under the buzzword-ridden surface, there are actually concepts out there that can be studied with more rigour.
+
+## State change
+
+The most fundamental thing in line-of-business software is a state change. State changes are the very reason we build such software. A system reacts on real-world events by changing its internal state. When you click "submit" an order entity is created. On the other end the state change of the system triggers an event in the real world. When the order is state is set to "payment succeeded" the parcel is dispatched to the courier.
+
+{{< figure src="/posts/software-system.png" title="Software system">}}
+
+## Chains of state changes
+
+A system as described above consists of a single data store and a single processing component. That is not how modern software systems are built. Regardless of your approach to architecture, the system is, sooner or later, going to consist of multiple components. One common thing that we can identify in such a system is chains of state changes, one triggering another.
+
+
+{{< figure src="/posts/chains.png" title="Chains of state changes">}}
+
+When you click "submit", the order and shipment entities are created. When the payment succeeds the order entity is modified. But that's not enough to get you need phone delivered to your door. When the order state is set to "payment succeeded", the shipment entity has to be modified too in order to unlock the delivery. One state change triggers another state change, possibly in another part of the system.
+
+## Distributed business processes
+
+We call these chains of state changes triggering one another distributed business processes. You can find them in most line-of-business systems. They are interesting because they are not specific to the approach to architecture you may have. 
+
+Looking at the distributed business processes in more formal way, we can describe them using following statements:
+- if the initial change has been persisted, the follow-up change will eventually be persisted too
+- for each instance of initial state change, the follow-up state change has to occur exactly once
+- the process that executed the follow-up state change has to see the initial state change as done
```

Comment:
I think that we need to introduce an outside observer to clarify this.  I don't think it's the follow-up process that needs to see the initial state changes but an external observer e.g. UI user. In other words, we want causal states to change visibility to prevent confusion when looking at the system externally.